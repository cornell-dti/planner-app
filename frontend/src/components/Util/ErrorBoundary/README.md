# [`ErrorBoundary`](./index.tsx) Documentation

This is the component that should exists at the top-level, wrapped all components. In this way,
it can provide error recovery support.

Currently, it only displays a funny broken Samwise bear and error reporting link when some error
occurs.
