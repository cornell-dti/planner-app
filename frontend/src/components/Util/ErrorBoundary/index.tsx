import React, { ReactElement } from 'react';
import FrontendCrash from '../../../assets/error/frontendCrash.png';

type Props = { readonly children: ReactElement };

type State = { readonly hasError: boolean };

export default class ErrorBoundary extends React.PureComponent<Props, State> {
  public readonly state: State = { hasError: false };

  public static getDerivedStateFromError(): State {
    return { hasError: true };
  }

  private ignoreError = (): void => this.setState({ hasError: false });

  public componentDidCatch(error: Error | null, info: object): void {
    // eslint-disable-next-line no-console
    console.log({ error, info }); // necessary for error logging!
  }

  public render(): ReactElement {
    const { children } = this.props;
    const { hasError } = this.state;
    if (hasError) {
      return (
        <div>
          <img
            src={FrontendCrash}
            alt="Uh oh - we broke something"
            style={{
              width: '100vw',
              height: '100vh',
              objectFit: 'cover',
              objectPosition: 'center',
              position: 'absolute',
              zIndex: -1,
              left: 0,
              top: 0,
            }}
          />
          <div style={{
            position: 'fixed',
            bottom: 0,
            right: 0,
            padding: '10px',
            textAlign: 'right',
            fontSize: '150%',
          }}
          >
            <p>
              Come yell at us:&nbsp;
              <a
                style={{ color: 'white', textDecoration: 'underline' }}
                href="https://goo.gl/forms/PZUZ1Ze6kN82EmcD2"
              >
                  Send Feedback
              </a>
            </p>
            <button type="button" onClick={this.ignoreError}>
              Click me to refresh the page. It may help.
            </button>
          </div>
        </div>
      );
    }
    return children;
  }
}
