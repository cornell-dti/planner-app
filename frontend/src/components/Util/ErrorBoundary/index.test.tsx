import React, { ReactElement } from 'react';
import renderer from 'react-test-renderer';
import ErrorBoundary from '.';

const BrokenComponent = (): ReactElement => {
  throw new Error('Ah');
};

it('ErrorBoundary can catch error and render', () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore: we want to simulate production error page.
  process.env.NODE_ENV = 'production';
  const tree = renderer.create(
    <ErrorBoundary><BrokenComponent /></ErrorBoundary>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
