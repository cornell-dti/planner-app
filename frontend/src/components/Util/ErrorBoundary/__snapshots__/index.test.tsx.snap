// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary can catch error and render 1`] = `
<div>
  <img
    alt="Uh oh - we broke something"
    src="/error/frontendCrash.png"
    style={
      Object {
        "height": "100vh",
        "left": 0,
        "objectFit": "cover",
        "objectPosition": "center",
        "position": "absolute",
        "top": 0,
        "width": "100vw",
        "zIndex": -1,
      }
    }
  />
  <div
    style={
      Object {
        "bottom": 0,
        "fontSize": "150%",
        "padding": "10px",
        "position": "fixed",
        "right": 0,
        "textAlign": "right",
      }
    }
  >
    <p>
      Come yell at us:Â 
      <a
        href="https://goo.gl/forms/PZUZ1Ze6kN82EmcD2"
        style={
          Object {
            "color": "white",
            "textDecoration": "underline",
          }
        }
      >
        Send Feedback
      </a>
    </p>
    <button
      onClick={[Function]}
      type="button"
    >
      Click me to refresh the page. It may help.
    </button>
  </div>
</div>
`;
